blockmanager.go

```diff
@@ -2400,255 +2409,274 @@ func (b *blockManager) handleHeadersMsg(hmsg *headersMsg) {
 		prevNodeEl := b.headerList.Back()
 		if prevNodeEl == nil {
 			log.Warnf("Header list does not contain a previous" +
-				"element as expected -- disconnecting peer")
-			hmsg.peer.Disconnect()
+				"element as expected -- disconnecting peer, if involved")
+			if hmsg.peer != nil {
+				hmsg.peer.Disconnect()
+			}
 			return
 		}
 
-		// Ensure the header properly connects to the previous one,
-		// that the proof of work is good, and that the header's
-		// timestamp isn't too far in the future, and add it to the
-		// list of headers.
 		node := headerlist.Node{Header: *blockHeader}
 		prevNode := prevNodeEl
 		prevHash := prevNode.Header.BlockHash()
-		if prevHash.IsEqual(&blockHeader.PrevBlock) {
```

Comment:
I'd prefer to minimize the set of changes to this area, as it's critical for operation of the daemon. IMO the existing logic can be simplified a ton, but all we really need from here is `checkHeaderSanity`.

Reply:
I think that was what I did initially but the last review talked about duplicating the `handle headers` functionality, so I updated the PR to prevent that.

Reply:
Are you suggesting the blockmanager and sideloader use the `handleheaders` function and then I refactor to allow that ?